src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -823,7 +834,7 @@ bool LibraryCallKit::inline_vector_mem_operation(bool is_store) {
 
   const TypeInstPtr* vbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, vbox_klass);
 
-  if (can_access_non_heap) {
+  if (mixed_access) {
```

Comment:
I suggest the following shapes for the checks:
```
const bool in_native = (TypePtr::NULL_PTR == base_type);            // base always null
const bool in_heap   = !TypePtr::NULL_PTR->higher_equal(base_type); // base never null

const bool is_mixed_access = !in_heap && !in_native;

const bool is_mismatched_access = in_heap && (addr_type->isa_aryptr() == NULL);

const bool needs_cpu_membar = is_mixed_access || is_mismatched_access;
```

Reply:
Thank you, fixed, we were not sure if mismatched access should be there.